worlds/sm64ex/Rules.py

```diff
@@ -1,67 +1,96 @@
+            randomized_level_to_paintings[Entrances.BOB_OMB_BATTLEFIELD] = first_course
+
     if world.AreaRandomizer[player].value == 2:  # Randomize Secrets as well
         randomized_level_to_secrets = shuffle_dict_keys(world,sm64_level_to_secrets)
-    randomized_entrances = { **randomized_level_to_paintings, **randomized_level_to_secrets }
+    randomized_entrances = {**randomized_level_to_paintings, **randomized_level_to_secrets}
     if world.AreaRandomizer[player].value == 3:  # Randomize Courses and Secrets in one pool
-        randomized_entrances = shuffle_dict_keys(world,randomized_entrances)
+        randomized_entrances = shuffle_dict_keys(world, randomized_entrances)
         # Guarantee first entrance is a course
-        swapdict = { entrance: level for (level,entrance) in randomized_entrances }
-        if randomized_entrances[91] not in sm64_paintings_to_level.keys(): # Unlucky :C (91 -> BoB Entrance)
-            rand = world.random.choice(sm64_paintings_to_level.values())
-            randomized_entrances[91], randomized_entrances[swapdict[rand]] = rand, randomized_entrances[91]
-            swapdict[rand] = randomized_entrances[91]
-        swapdict.pop("Bob-omb Battlefield")
+        swapdict = {entrance: level for (level, entrance) in randomized_entrances.items()}
+        valid_first_courses = valid_move_randomizer_start_courses if move_rando_bitvec > 0 else list(sm64_paintings_to_level.keys())
+        first_course = randomized_entrances[Entrances.BOB_OMB_BATTLEFIELD]
+        if first_course not in valid_first_courses:  # Unlucky :C
+            rand = world.random.choice(valid_first_courses)
+            rand_entrance = swapdict[rand]
+            randomized_entrances[Entrances.BOB_OMB_BATTLEFIELD], randomized_entrances[rand_entrance] = rand, first_course
+            swapdict.pop(rand)
+            swapdict[first_course] = rand_entrance
         # Guarantee COTMC is not mapped to HMC, cuz thats impossible
-        fix_reg(randomized_entrances, "Cavern of the Metal Cap", "Hazy Maze Cave", swapdict, world)
+        fix_reg(randomized_entrances, Entrances.CAVERN_OF_THE_METAL_CAP, "Hazy Maze Cave", swapdict, world)
         # Guarantee BITFS is not mapped to DDD
-        fix_reg(randomized_entrances, "Bowser in the Fire Sea", "Dire, Dire Docks", swapdict, world)
-        if randomized_entrances[191] == "Hazy Maze Cave": # If BITFS is mapped to HMC...
-            fix_reg(randomized_entrances, "Cavern of the Metal Cap", "Dire, Dire Docks", swapdict, world) # ... then dont allow COTMC to be mapped to DDD
+        fix_reg(randomized_entrances, Entrances.BOWSER_IN_THE_FIRE_SEA, "Dire, Dire Docks", swapdict, world)
+        if randomized_entrances[Entrances.BOWSER_IN_THE_FIRE_SEA] == "Hazy Maze Cave": # If BITFS is mapped to HMC...
+            fix_reg(randomized_entrances, Entrances.CAVERN_OF_THE_METAL_CAP, "Dire, Dire Docks", swapdict, world, {"Hazy Maze Cave"}) # ... then dont allow COTMC to be mapped to DDD
```

Comment:
This can be simplified using `invalid_regions`, if you remove the `destination` parameter
```
fix_reg for BitFS
fix_reg for COTMC (invalid regions both HMC and DDD)
```
So only 2 calls to fix_reg